keel-titus-plugin/src/main/kotlin/com/netflix/spinnaker/keel/api/titus/cluster/TitusClusterHandler.kt

```diff
@@ -112,31 +116,165 @@ class TitusClusterHandler(
     resourceDiff: ResourceDiff<Map<String, TitusServerGroup>>
   ): List<Task> =
     coroutineScope {
-      resourceDiff
+      val diffs = resourceDiff
         .toIndividualDiffs()
         .filter { diff -> diff.hasChanges() }
-        .map { diff ->
-          val desired = diff.desired
-          val job = when {
-            diff.isCapacityOnly() -> diff.resizeServerGroupJob()
-            else -> diff.upsertServerGroupJob() + resource.spec.deployWith.toOrcaJobProperties()
-          }
 
-          log.info("Upserting server group using task: {}", job)
+      val deferred: MutableList<Deferred<Task>> = mutableListOf()
+      val modifyDiffs = diffs.filter { it.isCapacityOnly() }
+      val createDiffs = diffs - modifyDiffs
 
-          async {
-            taskLauncher.submitJobToOrca(
-              resource = resource,
-              description = "Upsert server group ${desired.moniker.name} in ${desired.location.account}/${desired.location.region}",
-              correlationId = "${resource.id}:${desired.location.region}",
-              job = job
-            )
-          }
+      if (modifyDiffs.isNotEmpty()) {
+        deferred.addAll(
+          modifyInPlace(resource, modifyDiffs))
+      }
+
+      if (resource.spec.deployWith.isStaggered && createDiffs.isNotEmpty()) {
+        val tasks = upsertStaggered(resource, createDiffs)
+        return@coroutineScope tasks + deferred.map { it.await() }
+      }
+
+      deferred.addAll(
+        upsertUnstaggered(resource, createDiffs))
+
+      return@coroutineScope deferred.map { it.await() }
+    }
+
+  private suspend fun modifyInPlace(
+    resource: Resource<TitusClusterSpec>,
+    diffs: List<ResourceDiff<TitusServerGroup>>
+  ): List<Deferred<Task>> =
+    coroutineScope {
+      diffs.map { diff ->
+        val job = diff.resizeServerGroupJob()
+
+        log.info("Modifying server group in-place using task: $job")
+
+        async {
+          taskLauncher.submitJobToOrca(
+            resource = resource,
+            description = "Upsert server group ${diff.desired.moniker.name} in " +
+              "${diff.desired.location.account}/${diff.desired.location.region}",
+            correlationId = "${resource.id}:${diff.desired.location.region}",
+            job = job
+          )
+        }
+      }
+    }
+
+  private suspend fun upsertStaggered(
```

Comment:
Do you think any of this could be generalized or shared between any resource that wants to do a staggered rollout? This looks similar to the code in the ec2 cluster handler.

Reply:
I'd love suggestions or ideas!

Reply:
What I'm thinking is sort of along the lines of what I did for the Docker image resolver ([here](https://github.com/spinnaker/keel/blob/master/keel-docker/src/main/kotlin/com/netflix/spinnaker/keel/docker/DockerImageResolver.kt)). 

So, maybe we have a base class in `keel-plugin` that helps orchestrate things. Each handler that wanted to use it could implement some functions, like a function to get `isStaggered` from their own spec type.

Then there could be a function that takes the `create` diffs and the `modify` diffs, submits things to orca, and returns the task list?

Reply:
Interesting, I'll close this PR.